File: app/lib/models/item.dart
/// 物品模型类，用于表示系统中的物品信息
class Item {
  /// 物品唯一标识符
  final int id;
  /// 物品名称
  final String name;
  /// 物品描述（可选）
  final String? description;
  /// 分类ID
  final int categoryId;
  /// 位置ID
  final int locationId;
  /// 数量
  final int quantity;
  /// 单位
  final String unit;
  /// 过期日期（可选）
  final DateTime? expiryDate;
  /// 购买日期（可选）
  final DateTime? purchaseDate;
  /// 价格（可选）
  final double? price;
  /// 家庭ID
  final int familyId;
  /// 创建者ID
  final int createdBy;
  /// 创建时间
  final DateTime createdAt;
  /// 更新时间
  final DateTime updatedAt;

  /// 构造函数
  Item({
    required this.id,
    required this.name,
    this.description,
    required this.categoryId,
    required this.locationId,
    required this.quantity,
    required this.unit,
    this.expiryDate,
    this.purchaseDate,
    this.price,
    required this.familyId,
    required this.createdBy,
    required this.createdAt,
    required this.updatedAt,
  });

  /// 从JSON数据创建Item实例
  /// [json] JSON格式的物品数据
  factory Item.fromJson(Map<String, dynamic> json) {
    return Item(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      categoryId: json['category_id'],
      locationId: json['location_id'],
      quantity: json['quantity'],
      unit: json['unit'],
      expiryDate: json['expiry_date'] != null ? DateTime.parse(json['expiry_date']) : null,
      purchaseDate: json['purchase_date'] != null ? DateTime.parse(json['purchase_date']) : null,
      price: json['price'] != null ? double.parse(json['price'].toString()) : null,
      familyId: json['family_id'],
      createdBy: json['created_by'],
      createdAt: DateTime.parse(json['created_at']),
      updatedAt: DateTime.parse(json['updated_at']),
    );
  }

  /// 将Item实例转换为JSON格式
  /// 返回JSON格式的物品数据
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'category_id': categoryId,
      'location_id': locationId,
      'quantity': quantity,
      'unit': unit,
      'expiry_date': expiryDate?.toIso8601String(),
      'purchase_date': purchaseDate?.toIso8601String(),
      'price': price,
      'family_id': familyId,
      'created_by': createdBy,
      'created_at': createdAt.toIso8601String(),
      'updated_at': updatedAt.toIso8601String(),
    };
  }
}

/// 分类模型类，用于表示物品的分类信息
class Category {
  /// 分类唯一标识符
  final int id;
  /// 分类名称
  final String name;
  /// 分类图标（可选）
  final String? icon;
  /// 分类颜色（可选）
  final String? color;
  /// 家庭ID
  final int familyId;
  /// 创建时间
  final DateTime createdAt;

  /// 构造函数
  Category({
    required this.id,
    required this.name,
    this.icon,
    this.color,
    required this.familyId,
    required this.createdAt,
  });

  /// 从JSON数据创建Category实例
  /// [json] JSON格式的分类数据
  factory Category.fromJson(Map<String, dynamic> json) {
    return Category(
      id: json['id'],
      name: json['name'],
      icon: json['icon'],
      color: json['color'],
      familyId: json['family_id'] ?? json['family'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  /// 将Category实例转换为JSON格式
  /// 返回JSON格式的分类数据
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'icon': icon,
      'color': color,
      'family_id': familyId,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

/// 位置模型类，用于表示物品的存放位置信息
class Location {
  /// 位置唯一标识符
  final int id;
  /// 位置名称
  final String name;
  /// 位置描述（可选）
  final String? description;
  /// 父位置ID（可选）
  final int? parent;
  /// 父位置名称（可选）
  final String? parentName;
  /// 家庭ID
  final int familyId;
  /// 创建时间
  final DateTime createdAt;

  /// 构造函数
  Location({
    required this.id,
    required this.name,
    this.description,
    this.parent,
    this.parentName,
    required this.familyId,
    required this.createdAt,
  });

  /// 从JSON数据创建Location实例
  /// [json] JSON格式的位置数据
  factory Location.fromJson(Map<String, dynamic> json) {
    return Location(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      parent: json['parent'],
      parentName: json['parent_name'],
      familyId: json['family_id'] ?? json['family'],
      createdAt: DateTime.parse(json['created_at']),
    );
  }

  /// 将Location实例转换为JSON格式
  /// 返回JSON格式的位置数据
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'parent': parent,
      'parent_name': parentName,
      'family_id': familyId,
      'created_at': createdAt.toIso8601String(),
    };
  }
}

File: HomeWareServer/app/models/item.py
"""
物品模型模块
定义 Item 和 ItemImage 模型
"""
from sqlalchemy import Boolean, Column, Date, DateTime, ForeignKey, Integer, Numeric, String, Text
from sqlalchemy.orm import relationship

from app.core.database import Base
from app.models.base import BaseMixin


class Item(Base, BaseMixin):
    """物品模型"""
    
    __tablename__ = "items"
    
    name = Column(String(100), nullable=False, comment="物品名称")
    brand = Column(String(50), nullable=True, comment="品牌")
    specification = Column(String(100), nullable=True, comment="规格")
    barcode = Column(String(50), nullable=True, comment="条码")
    category_id = Column(Integer, ForeignKey("categories.id"), nullable=False, comment="分类ID")
    location_id = Column(Integer, ForeignKey("locations.id"), nullable=True, comment="位置ID")
    container_name = Column(String(50), nullable=True, comment="容器名（收纳箱/药盒等）")
    family_id = Column(Integer, ForeignKey("families.id"), nullable=False, comment="家庭ID")
    
    # 价格相关
    purchase_price = Column(Numeric(10, 2), nullable=True, comment="购买单价")
    total_price = Column(Numeric(10, 2), nullable=True, comment="总价")
    purchase_quantity = Column(Integer, default=1, comment="购买数量（包装数）")
    package_unit = Column(String(10), nullable=True, comment="包装单位（盒/箱/提）")
    package_quantity = Column(Integer, default=1, comment="一包装含多少基本单位")
    current_quantity = Column(Numeric(10, 2), default=1, comment="当前数量（总基本单位）")
    unit = Column(String(10), default="件", comment="基本单位（片/瓶/个）")
    safety_stock = Column(Numeric(10, 2), default=1, comment="安全库存")
    
    # 日期相关
    purchase_date = Column(Date, nullable=True, comment="购买日期")
    purchase_channel = Column(String(50), nullable=True, comment="购买渠道")
    production_date = Column(Date, nullable=True, comment="生产日期")
    expiry_date = Column(Date, nullable=True, comment="过期日期")
    shelf_life_days = Column(Integer, nullable=True, comment="保质期天数")
    opened_date = Column(Date, nullable=True, comment="开封日期")
    after_open_days = Column(Integer, nullable=True, comment="开封后保质期天数")
    warranty_date = Column(Date, nullable=True, comment="质保日期")
    
    # 提醒设置
    expiry_alert_days = Column(Integer, default=3, comment="过期提醒天数")
    stock_alert = Column(Boolean, default=True, comment="是否开启库存提醒")
    
    # 其他
    notes = Column(Text, nullable=True, comment="备注")
    status = Column(Integer, default=0, comment="状态(0使用中/1用完/2过期/3丢弃)")
    avg_daily_consumption = Column(Numeric(10, 4), nullable=True, comment="日均消耗量")
    predicted_empty_date = Column(Date, nullable=True, comment="预计用完日期")
    created_by = Column(Integer, ForeignKey("users.id"), nullable=False, comment="创建人ID")
    deleted_at = Column(DateTime, nullable=True, comment="删除时间(软删除)")
    
    # 关系定义
    category = relationship("Category")
    location = relationship("Location")
    family = relationship("Family")
    created_by_user = relationship("User")
    images = relationship("ItemImage", back_populates="item", cascade="all, delete-orphan")
    usage_records = relationship("UsageRecord", back_populates="item")


class ItemImage(Base, BaseMixin):
    """物品图片模型"""
    
    __tablename__ = "item_images"
    
    item_id = Column(Integer, ForeignKey("items.id"), nullable=False, comment="物品ID")
    url = Column(String(500), nullable=False, comment="图片路径/URL")
    sort_order = Column(Integer, default=0, comment="排序序号")
    
    # 关系定义
    item = relationship("Item", back_populates="images")